### PR TITLE
Initial introduction of vector guess

### PR DESCRIPTION
**Pull request by andrewram4287**

### CODE REVIEW

File: gmm/tensor.py [hunk 1]
```diff
@@ -83,6 +86,25 @@ def vector_to_cp_pt(vectorIn, rank: int, shape: tuple):
...
+
+    factortotal = (
+        factortotal
+        - (len(zflowTensor.coords["Cell"]) * ranknumb)
+        + (ranknumb * maxcluster)
+        + ((ranknumb**4) * (len(markerslist) ** 2))
+        + maxcluster
+    )
+
+    rand_vec = np.random.random(factortotal)
```
Comment (aarmey) [inline gmm/tensor.py:103]: Note this should be lognormal to ensure they're positive.

File: gmm/tensor.py [hunk 2]
```diff
@@ -143,3 +165,29 @@ def maxloglik_ptnnp(facVector, facInfo: tl.cp_tensor.CPTensor, zflowTensor: xa.D
     rebuildPrecision = (rebuildPrecision + np.swapaxes(rebuildPrecision, 1, 2)) / 2.0  # Enforce symmetry
     # Creating function that we want to minimize
     return -comparingGMMjax(zflowTensor.to_numpy(), rebuildMeans, rebuildPrecision, rebuildnk)
+
+
+def minimize_func(totalVector, facInfo, zflowTensor, tMeans):
```
Comment (aarmey) [inline gmm/tensor.py:170]: Docstrings for both new functions.

**Review by aarmey — COMMENTED** [review]
Ok, clean it up like this, then I'll take a look at removing facInfo and tMeans from minimize_func. I think it should be possible but will taking some playing around.

**Comment by aarmey** [comment]
Note that probGMM gets its information for tMeans from somewhere, so this is where to look to get rid of the need for probGMM.

File: gmm/tensor.py [hunk 3]
```diff
@@ -143,3 +165,29 @@ def maxloglik_ptnnp(facVector, facInfo: tl.cp_tensor.CPTensor, zflowTensor: xa.D
...
+
+    rebuildCpFactors, _, ptNewCore = vector_to_cp_pt(opt.x[facInfo.shape[0] : :], facInfo.rank, facInfo.shape)
+    maximizedCpInfo = cp_normalize(rebuildCpFactors)
+    maximizedNK = opt.x[0 : facInfo.shape[0]]
+
+    cmpCol = [f"Cmp. {i}" for i in np.arange(1, facInfo.rank + 1)]
+
+    maximizedFactors = []
+    for ii, dd in enumerate(tMeans.dims):
+        maximizedFactors.append(pd.DataFrame(maximizedCpInfo.factors[ii], columns=cmpCol, index=tMeans.coords[dd]))
```
Comment (aarmey) [inline gmm/tensor.py:191]: I think these coords can be built from zFlowTensor instead.

File: gmm/tensor.py [hunk 4]
```diff
@@ -143,3 +165,29 @@ def maxloglik_ptnnp(facVector, facInfo: tl.cp_tensor.CPTensor, zflowTensor: xa.D
...
+    tl.set_backend("jax")
+
+    func = value_and_grad(maxloglik_ptnnp)
+
+    bnds = Bounds(np.zeros_like(totalVector), np.full_like(totalVector, np.inf), keep_feasible=True)
+    opt = minimize(func, totalVector, bounds=bnds, jac=True, method="L-BFGS-B", args=args, options={"iprint": 1, "maxiter": 500})
+
+    tl.set_backend("numpy")
+
+    rebuildCpFactors, _, ptNewCore = vector_to_cp_pt(opt.x[facInfo.shape[0] : :], facInfo.rank, facInfo.shape)
```
Comment (aarmey) [inline gmm/tensor.py:183]: You can provide the rank and shape to minimize_func, instead of facInfo.

**Comment by aarmey** [comment]
I've simplified what `minimize_func()` takes. Once there are docstrings and type annotations here this should be ready to merge.

**Comment by aarmey** [comment]
I've simplified what `minimize_func()` takes. Once there are docstrings and type annotations here this should be ready to merge.

**Comment by andrewram4287** [comment]
@aarmey Ok sounds good! Sorry for not helping out as much as I wanted to, I was going to work on the code but classes and hw had me busy today.

**Pull request merged**